runtime/queries/metricsview_comparison_toplist.go

```diff
@@ -545,7 +546,159 @@ func (q *MetricsViewComparisonToplist) buildMetricsComparisonTopListSQL(mv *runt
+			var sql string
+			var args []any
+			if !isTimeRangeNil(q.ComparisonTimeRange) {
+				sql, args, err = q.buildMetricsComparisonTopListSQL(q.MetricsView, olap.Dialect(), q.ResolvedMVSecurity)
+				if err != nil {
+					return fmt.Errorf("error building query: %w", err)
+				}
+			} else {
+				sql, args, err = q.buildMetricsTopListSQL(q.MetricsView, olap.Dialect(), q.ResolvedMVSecurity)
+				if err != nil {
+					return fmt.Errorf("error building query: %w", err)
+				}
+			}
+
+			filename := q.generateFilename(q.MetricsView)
+			if err := duckDBCopyExport(ctx, w, opts, sql, args, filename, olap, opts.Format); err != nil {
+				return err
+			}
+		} else {
+			if err := q.generalExport(ctx, rt, instanceID, w, opts, q.MetricsView); err != nil {
+				return err
+			}
+		}
+	case drivers.DialectDruid:
+		if err := q.generalExport(ctx, rt, instanceID, w, opts, q.MetricsView); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("not available for dialect '%s'", olap.Dialect())
+	}
+
+	return nil
+}
+
+func (q *MetricsViewComparisonToplist) generalExport(ctx context.Context, rt *runtime.Runtime, instanceID string, w io.Writer, opts *runtime.ExportOptions, mv *runtimev1.MetricsView) error {
+	err := q.Resolve(ctx, rt, instanceID, opts.Priority)
+	if err != nil {
+		return err
+	}
+
+	if opts.PreWriteHook != nil {
+		err = opts.PreWriteHook(q.generateFilename(mv))
+		if err != nil {
+			return err
+		}
+	}
+
+	var metaLen int
+	if !isTimeRangeNil(q.ComparisonTimeRange) {
+		metaLen = len(q.Result.Rows[0].MeasureValues) * 4
+	} else {
+		metaLen = len(q.Result.Rows[0].MeasureValues)
+	}
+	meta := make([]*runtimev1.MetricsViewColumn, metaLen+1)
+	meta[0] = &runtimev1.MetricsViewColumn{
+		Name: q.DimensionName,
+	}
+	if !isTimeRangeNil(q.ComparisonTimeRange) {
+		for i, m := range q.Result.Rows[0].MeasureValues {
+			meta[1+i*4] = &runtimev1.MetricsViewColumn{
+				Name: fmt.Sprintf("%s_base_value", m.MeasureName),
+			}
+			meta[2+i*4] = &runtimev1.MetricsViewColumn{
+				Name: fmt.Sprintf("%s_comparison_value", m.MeasureName),
+			}
+			meta[3+i*4] = &runtimev1.MetricsViewColumn{
+				Name: fmt.Sprintf("%s_delta_abs", m.MeasureName),
+			}
+			meta[4+i*4] = &runtimev1.MetricsViewColumn{
+				Name: fmt.Sprintf("%s_delta_rel", m.MeasureName),
+			}
+		}
+	} else {
+		for i, m := range q.Result.Rows[0].MeasureValues {
+			meta[1+i] = &runtimev1.MetricsViewColumn{
+				Name: fmt.Sprintf("%s_value", m.MeasureName),
+			}
+		}
+	}
+
+	data := make([]*structpb.Struct, len(q.Result.Rows))
+	for i, row := range q.Result.Rows {
+		data[i] = &structpb.Struct{
+			Fields: map[string]*structpb.Value{
+				q.DimensionName: {
+					Kind: &structpb.Value_StringValue{
+						StringValue: row.DimensionValue.GetStringValue(),
+					},
+				},
+			},
+		}
+		for _, m := range row.MeasureValues {
+			if !isTimeRangeNil(q.ComparisonTimeRange) {
```

Comment:
Given the number of iterations, maybe pre-compute this instead of calling every time?